src/commands/dev/audit/messages.ts

```diff
@@ -0,0 +1,478 @@
+              this.logger.trace(
+                `Try to add edge from message ref ${mesageRefNodeKey} to message ${bundleRefNode.name}:${key}`
+              );
+              // add an edge from the message reference to the message
+              this.graph.addEdge(mesageRefNodeKey, `${bundleRefNode.name}:${key}`);
+            }
+          }
+        });
+    });
+  }
+
+  private buildAuditResults(): void {
+    // find unused bundles
+    this.auditResults.unusedBundles = this.graph
+      .filterNodes((node, attrs) => attrs.type === 'bundle' && this.graph.inDegree(node) === 0)
+      .sort();
+
+    // find missing bundles
+    this.auditResults.missingBundles = this.graph
+      .filterNodes((node, attrs) => attrs.type === 'bundleReference' && this.graph.outDegree(node) === 0)
+      .map((node) => {
+        const bundleRefNode = this.graph.getNodeAttributes(node) as BundleRefNode;
+        const [File] = node.split(':');
+        return { Bundle: bundleRefNode.name, File, SourceVar: bundleRefNode.variable };
+      })
+      .sort((a, b) => {
+        const fileCompare = a.File.localeCompare(b.File);
+        const sourceVarCompare = a.SourceVar.localeCompare(b.SourceVar);
+        if (fileCompare === 0) {
+          return sourceVarCompare;
+        }
+        return fileCompare;
+      });
+
+    // find message references where are no outbound edges to messages
+    this.auditResults.missingMessages = this.graph
+      .filterNodes(
+        (node, attrs) =>
+          attrs.type === 'messageReference' &&
+          !this.graph.someOutboundNeighbor(node, (msgNode, msgAtrs) => {
+            return msgAtrs.type === 'message';
+          })
+      )
+      .map((key) => {
+        const bundleRef = this.graph.findInboundNeighbor(key, (node, attrs) => {
+          return attrs.type === 'bundleReference';
+        });
+        if (!bundleRef) {
+          throw new Error(`Unable to find bundle reference for ${key}`);
+        }
+        const bundle = this.graph.getNodeAttributes(bundleRef) as BundleRefNode;
+        const messageRef = this.graph.getNodeAttributes(key) as MessageRefNode;
+        const [File, SourceVar, Name] = key.split(':');
+        return { File, SourceVar, Name, IsLiteral: messageRef.isLiteral, Bundle: bundle.name };
+      })
+      .sort((a, b) => {
+        const fileCompare = a.File.localeCompare(b.File);
+        const nameCompare = a.Name.localeCompare(b.Name);
+        const sourceVarCompare = a.SourceVar.localeCompare(b.SourceVar);
+        if (fileCompare === 0) {
+          if (sourceVarCompare === 0) {
+            return nameCompare;
+          }
+          return sourceVarCompare;
+        }
+        return fileCompare;
+      });
+
+    const nonLiteralMessageBundleRefs = this.auditResults.missingMessages
+      .filter((m) => !m.IsLiteral)
+      .map((m) => m.Bundle)
+      .reduce((acc, bundle) => acc.add(bundle), new Set<string>());
```

Comment:
this might save another trip around the loop and avoid `reduce`.
```suggestion
    const nonLiteralMessageBundleRefs = new Set(this.auditResults.missingMessages
      .filter((m) => !m.IsLiteral)
      .map((m) => m.Bundle))
```